File: paoc/paoc.view.web/src/main/java/br/cefetmg/paoc/negocio/dao/admDAO.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package br.cefetmg.paoc.negocio.dao;

import br.cefetmg.paoc.database.Conexao;
import br.cefetmg.paoc.negocio.dto.Admin;
import br.cefetmg.paoc.negocio.excessoes.*;
import br.cefetmg.paoc.negocio.servico.persistenciaException;
import java.io.UnsupportedEncodingException;
import static java.lang.Integer.parseInt;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;

public class admDAO {

    public static List<Admin> listarAdmBD() throws SQLException {
        List<Admin> listaAdmin = new ArrayList<>();
        String sql = "SELECT * FROM administradores";
        Connection com = Conexao.getConnection();
        PreparedStatement stmt = com.prepareStatement(sql);
        ResultSet rs = stmt.executeQuery();

        while (rs.next()) {
            int id = Integer.parseInt(rs.getString("id"));
            String nome = rs.getString("nome");
            String email = rs.getString("email");
            String username = rs.getString("username");
            String senha = rs.getString("senha");

            Admin a = new Admin(nome, username, email, senha);
            listaAdmin.add(a);
        }
        if (listaAdmin.isEmpty() != true) {
            return listaAdmin;
        }
        return null;
    }

    public static String criptografarSenha(String senha) throws UnsupportedEncodingException, NoSuchAlgorithmException {
        MessageDigest algorithm = MessageDigest.getInstance("SHA-256");
        byte messageDigest[] = algorithm.digest(senha.getBytes("UTF-8"));

        StringBuilder hexString = new StringBuilder();
        for (byte b : messageDigest) {
            hexString.append(String.format("%02X", 0xFF & b));
        }
        String senhahex = hexString.toString();
        return senhahex;
    }

    public static Admin procurarAdmin(String username) throws SQLException {
        List<Admin> listaAdmins = listarAdmBD();

        if (listaAdmins != null) {
            for (Admin a : listaAdmins) {
                if (a.getUsername().equals(username)) {
                    return a;
                }
            }
        }
        return null;
    }

    public static void cadastrarAdmin(String nome, String email, String username, String senhaPura) throws negocioException, SQLException, UnsupportedEncodingException, NoSuchAlgorithmException {
        if (procurarAdmin(username) == null | listarAdmBD() == null) {
            try {
                String senha = criptografarSenha(senhaPura);
                Connection connection = Conexao.getConnection();
                String sql = "INSERT INTO administradores (nome, usuario, email, senha) VALUES (?, ?, ?, ?)";
                PreparedStatement pstmt = connection.prepareStatement(sql);
                pstmt.setString(1, nome);
                pstmt.setString(2, username);
                pstmt.setString(3, email);
                pstmt.setString(4, senha);
                pstmt.executeUpdate();
                System.out.println("Administrador cadastrado com sucesso!");
                pstmt.close();
                connection.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
    }

    public static void removerAdmin(String username) throws SQLException, negocioException {
        if (procurarAdmin(username) != null) {
            Admin a = procurarAdmin(username);
            String sql = "DELETE FROM administradores WHERE username='" + username + "'";
            Connection com = Conexao.getConnection();
            PreparedStatement pstmt = com.prepareStatement(sql);
            pstmt.execute();
            pstmt.close();
            com.close();
        }
    }

    public static Admin logarAdmin(String username, String senha) throws SQLException, UnsupportedEncodingException, NoSuchAlgorithmException, persistenciaException {
        Admin a = null;
        String sql = "SELECT * FROM administradores WHERE usuario='" + username + "' AND senha='" + senha + "'";
        System.out.println("Consultando no BD");
        Connection com = Conexao.getConnection();
        PreparedStatement stmt = com.prepareStatement(sql);
        ResultSet rs = stmt.executeQuery();
        if (rs.next()) {
            System.out.println("Tem usuário.");
            int id = Integer.parseInt(rs.getString("id"));
            String nome = rs.getString("nome");
            String email = rs.getString("email");
            String usuario = rs.getString("usuario");
            String senhaa = rs.getString("senha");
            System.out.println(nome);
            System.out.println(email);
            a = new Admin(nome, username, email, senhaa);

        } else {
            throw new persistenciaException("Usuario ou senha incorretos!");
        }
        return a;
    }
}

File: services/admDao.php
<?php

require_once 'Conexao.php';
require_once 'Admin.php';
require_once 'negocioException.php';
require_once 'persistenciaException.php';

class admDAO {

    public static function listarAdmBD() {
        $listaAdmin = array();
        $sql = "SELECT * FROM administradores";
        $com = Conexao::getConnection()->prepare($sql);
        $com->execute();

        while ($row = $com->fetch(PDO::FETCH_ASSOC)) {
            $id = (int)$row["id"];
            $nome = $row["nome"];
            $email = $row["email"];
            $username = $row["usuario"];
            $senha = $row["senha"];
            $permissao = $row["permissao"];

            $a = new Admin($nome, $username, $email, $senha, $permissao);
            $listaAdmin[] = $a;
        }
        
        if (!empty($listaAdmin)) {
            return $listaAdmin;
        }
        
        return null;
    }

    public static function criptografarSenha($senha) {
        return hash('sha256', $senha);
    }

    public static function procurarAdmin($username) {
        $listaAdmins = self::listarAdmBD();

        if (!empty($listaAdmins)) {
            foreach ($listaAdmins as $a) {
                if ($a->getUsername() === $username) {
                    return $a;
                }
            }
        }

        return null;
    }

    public static function cadastrarAdmin($nome, $email, $username, $senhaPura, $permissao) {
        try {
            if (self::procurarAdmin($username) === null || self::listarAdmBD() === null) {
                $senha = $senhaPura;
                $connection = Conexao::getConnection();
                $sql = "INSERT INTO administradores (nome, usuario, email, senha, permissao) VALUES (?, ?, ?, ?, ?)";
                $pstmt = $connection->prepare($sql);
                $pstmt->bindParam(1, $nome, PDO::PARAM_STR);
                $pstmt->bindParam(2, $username, PDO::PARAM_STR);
                $pstmt->bindParam(3, $email, PDO::PARAM_STR);
                $pstmt->bindParam(4, $senha, PDO::PARAM_STR);
                $pstmt->bindParam(5, $permissao, PDO::PARAM_STR);
                $pstmt->execute();
                echo "Administrador cadastrado com sucesso!";
                $pstmt->closeCursor();
            }
        } catch (Exception $e) {
            $e->getMessage();
        }
    }

    public static function removerAdmin($username) {
        try {
            $admin = self::procurarAdmin($username);
            if ($admin !== null) {
                $sql = "DELETE FROM administradores WHERE usuario = ?";
                $com = Conexao::getConnection()->prepare($sql);
                $com->bindParam(1, $username, PDO::PARAM_STR);
                $com->execute();
                echo "Admin removido com sucesso!";
            }
        } catch (Exception $e) {
            $e->getMessage();
        }
    }

    public static function logarAdmin($username, $senha) {
        try {
            $sql = "SELECT * FROM administradores WHERE usuario = ?";
            $com = Conexao::getConnection()->prepare($sql);
            $com->bindParam(1, $username, PDO::PARAM_STR);
            $com->execute();
            $row = $com->fetch(PDO::FETCH_ASSOC);
    
            if ($row) {
                // Verificar a senha criptografada no banco de dados
                if ($row["senha"] === $senha) {
                    $nome = $row["nome"];
                    $email = $row["email"];
                    $usuario = $row["usuario"];
                    $senhaa = $row["senha"];
                    $permissao = $row["permissao"];
                    $a = new Admin($nome, $usuario, $email, $senhaa, $permissao);
                    return $a;
                } else {
                    throw new persistenciaException("Senha incorreta!");
                }
            } else {
                throw new persistenciaException("Usuário não encontrado!");
            }
        } catch (Exception $e) {
            throw new persistenciaException("Erro ao realizar login: " . $e->getMessage());
        }
    }
}
